service/worker/pernamespaceworker_test.go

```diff
@@ -498,6 +501,51 @@ func (s *perNsWorkerManagerSuite) TestStartWorkerError() {
 	cli2.EXPECT().Close()
 }
 
+func (s *perNsWorkerManagerSuite) TestRateLimit() {
+	mockLimiter := quotas.NewMockRateLimiter(s.controller)
+	s.manager.startLimiter = mockLimiter
+
+	// try to start 100 workers
+	// rate limiter will allow 10, then 10 more after 0.2s, then no more for 10s
+	for i := 0; i < 100; i++ {
+		res := quotas.NewMockReservation(s.controller)
+		mockLimiter.EXPECT().Reserve().Return(res)
+		switch i % 10 {
+		case 2:
+			res.EXPECT().Delay().Return(0 * time.Millisecond)
+		case 6:
+			res.EXPECT().Delay().Return(200 * time.Millisecond)
+		default:
+			res.EXPECT().Delay().Return(10 * time.Second)
+		}
+	}
+
+	s.cmp1.EXPECT().DedicatedWorkerOptions(gomock.Any()).Return(&workercommon.PerNSDedicatedWorkerOptions{Enabled: true}).AnyTimes()
+	s.cmp2.EXPECT().DedicatedWorkerOptions(gomock.Any()).Return(&workercommon.PerNSDedicatedWorkerOptions{Enabled: false}).AnyTimes()
+	s.cmp1.EXPECT().Register(gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
+	s.serviceResolver.EXPECT().LookupN(gomock.Any(), 1).Return([]membership.HostInfo{membership.NewHostInfoFromAddress("self")}).AnyTimes()
+	cli := mocksdk.NewMockClient(s.controller)
+	wkr := mocksdk.NewMockWorker(s.controller)
+	s.cfactory.EXPECT().NewClient(gomock.Any()).Return(cli).AnyTimes()
+	s.cfactory.EXPECT().NewWorker(gomock.Any(), gomock.Any(), gomock.Any()).Return(wkr).AnyTimes()
+	var starts atomic.Int32
+	wkr.EXPECT().Start().Do(func() { starts.Add(1) }).AnyTimes()
+
+	for i := 0; i < 100; i++ {
+		ns := testns(fmt.Sprintf("test-%d", i), enumspb.NAMESPACE_STATE_REGISTERED)
+		s.manager.namespaceCallback(ns, false)
+	}
+
+	time.Sleep(100 * time.Millisecond)
+	s.Equal(10, int(starts.Load()), "should be 10 started immediately")
+
+	time.Sleep(200 * time.Millisecond)
+	s.Equal(20, int(starts.Load()), "should be 20 started later")
```

Comment:
This makes the tests for this package go from 0.843s to 1.151s. Unit tests for the whole server take about 3½ minutes vs over 10 minutes for the largest slice of functional tests, so unit tests would have to triple in length to have an effect on CI. I appreciate the concern but I don't think it's worth focusing much effort on this type of thing right now.

It's testing a rate limit, which is fundamentally time-based. To do better (I did try) we'd have to refactor `perNamespaceWorkerManager` to use a fake clock throughout. That's a good thing to do and I agree we should, but I didn't want to get into that sort of refactoring right now. I can leave a TODO if you want

Reply:
> Unit tests for the whole server take about 3½ minutes vs over 10 minutes for the largest slice of functional tests

I don't think that's justifying having slow unit tests though. It just discourages users from running unit test frequently during local development.
My point is that I would rather take steps to prevent the situation from getting worse than continue this pattern.
I think my suggestion to use a notification channel and lower delays serves that purpose and eliminates the chances of flakiness.

I won't block the PR on this, leaving it up to you to make a call.

Reply:
I rearranged it to use a channel and take 50ms. I'm hesitant to decrease that or it might get flaky. There's no way to make it totally non-flaky without refactoring the timer-related stuff, and that will have to wait for another PR.